### ISSUE TITLE
fatal error C1060: compiler is out of heap space

### ISSUE THREAD

**Issue by ohmtech-rdi** [appveyor/ci#742]
Hi,

I have a problem when trying to build my project with Visual Studio 2015.

### Description

I'm using **MSBuild 14.0** and the **2015 compiler** and encounter the following error :
```
 fatal error C1060: compiler is out of heap space
 ```
quite randomly.
After that, random errors like the following will appear, but the one described upper is always the first one.
```
fatal error C1090: PDB API call failed, error code '24':
Internal Compiler Error in C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64_x86\CL.exe.
cl : Command line error D8040: error creating or communicating with child process
```

Using **MSBuild 12.0** and the **2013 compiler** on the same build worker image is running **perfectly**.

The visual 2013 & 2015 solutions are generated using ```gyp```.

The project itself is on GitHub and is private, so I can't share here the code I'm using. But I don't think this is a big issue for the following reasons :
 - It's C++ code, it is quite modern C++ but not heavily templatized
 - I can reproduce the error on a local VM with interesting results described below

### Findings

I can reproduce the problem on a local VM I have set up with Windows 10 and Visual 2015. If I set the VM memory to 4 GB (the same as the build worker image), then it will show the same behaviour.
Also I know it **does not come from the C++ code** itself, as I can run MSBuild several times and after some point (at most 3 consecutive builds) have my project correctly built and linked.

The behaviour upper is the same with update 1 or update 2 of MSVS2015.

**I'm not using precompiled headers** (this is what shows up mostly when searching for this error on google/stackoverflow)

### Question

I'm quite surprised as it seems I'm the only one facing this error. But maybe others found a working and acceptable workaround for this problem?

Any information on that matter would be greatly appreciated.

Thanks,

Raphael

**Comment by mwhouser** [comment]
I used to see this error when compiling some `.cpp` files that, when including many header files, would become very large. Essentially, the compiler is running out of memory trying to allocate space for your very large file.

https://msdn.microsoft.com/en-us/library/yz7kx3y2(v=vs.140).aspx

Sometimes, this issue would occur on one developer's PC, but not others. But when it did occur on a PC, it was quite consistent.

To fix, one option that worked for me was to split the very large `.cpp` file into multiple files. But this did not always work.

Other times, I was able to solve this problem by using the `/Zm` compiler option to set/increase the memory limit.

**Comment by ohmtech-rdi** [comment]
Hi,

First, thanks for your answer.

My `.cpp` files are quite small (38 KB for the biggest, almost all others are below 20 KB), and my headers use forward declaration extensively.

However, since I'm testing the library with a configuration matrix, I end up with quite a number of projects in the solution (with different precompiler flags), and I guess the total number of files to compile might be around 1500. Each project has around 100 code unit `.cpp` files.

Reading the `/Zm` documentation :
https://msdn.microsoft.com/en-us/library/bdscwf1c.aspx
says that it is used to "Determines the amount of memory that the compiler allocates to construct precompiled headers"

I'm not using precompiled headers. Do you think that the documentation is actually faulty and there is an influence even in the case where I wouldn't be using precompiled headers? (or an misunderstanding of me of what could be precompiled headers in that context?)

**Comment by mwhouser** [comment]
My experience is that the issue occurs (and is fixed) regardless of precompiled header use. So give it a try (wouldn't hurt). I was able to do `/Zm100` and it solved my issue.  Also, my understanding is that the memory allocation is a per-file thing, not a per-project thing. So it's the size of an individual `.cpp` file.

**Comment by kelleyma49** [comment]
We see this issue often and it's intermittent.  We usually see it in the middle of a full rebuild.  I believe it's intermittent because multiple files may be passed on the command line at once due to a multiprocessor build.

We've found adding RAM and increasing virtual memory seems to help diminish this issue.

**Comment by ohmtech-rdi** [comment]
@mwhouser Thanks for your advice. Meantime I've tried with `/Zm100`, `/Zm200` and `/Zm10` and it seems that the latter "makes it work". I'm not super confident, as I don't understand what I'm doing, and the MSDN documentation is completely misleading on this one.

@kelleyma49 Yes, this is exactly what I was seeing. Usually it would always fail around the 500th code unit. I can't increase the RAM of the VM myself as I'm running on the AppVeyor CI system.

**Comment by ohmtech-rdi** [comment]
@mwhouser Ah no actually it's not working reliably. Relaunched the same build to test and it had the same issue.

**Comment by ohmtech-rdi** [comment]
Thanks everyone for your input. It helped to understand more the problem, and we found a solution, that I'm going to expose here, as it was a misuse of `MSBuild`/`CL` arguments, that could go unnoticed on a typical developer workstation but will most probably makes problem on a limited memory VM configuration.

We are creating on the fly a MSVS solution using `gyp`. Then we use `MSBuild` to build the solution.

We were using the compiler option `/MP` for the solution as well as the argument `/maxcpucount` for `MSBuild`.
In this configuration, and on AppVeyor VM, `MSBuild` will "fork" 4 times and `CL` will also "fork" 4 times, resulting in 16 compilers running in parallel. Some code units require more heap memory than others, and so we could run in the situation where we would exhaust the available memory as each compiler process have _on average_ a bit less than 200 MB to run.

Removing `/maxcpucount` or `/MP` solves the problem.
As far as build time is concerned, it **seems** that keeping `/maxcpucount` and removing `/MP` gives better overall build duration (The time fluctuation is small enough to be considered not significant though, and could be related to overall AppVeyor platform charge).

As a matter of fact, this behaviour is described in details on a MSDN page here :
https://msdn.microsoft.com/en-us/library/bb385193.aspx
(bottom of the page, in the "The MSBUILD.exe Tool" section)

Thanks,

Raphael

**Comment by FeodorFitsner** [comment]
Thanks for sharing the solution!

**Comment by vinipsmaker** [comment]
Hi, I'm trying to use AppVeyor, but my project keeps stumbling on C1060: https://ci.appveyor.com/project/vinipsmaker/emilua/builds/46389030

I disabled pre-compiled headers and parallel build and I keep stumbling on the issue.

I tried to modify `/Zm` to `100` and `0` and neither option works.

Microsoft docs list a few options to solve the issue: https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1060?view=msvc-170

`/Zm` was one of the hints to solve the issue (which — again — didn't work).

Next on the list is to increase the size of the Windows swap-file. How do I do that? I can't find docs on how to do that for Appveyor.

I don't know what else I can do to solve the issue.
